### PR TITLE
Add unique ID for tokens, and associated metadata

### PR DESCRIPTION
**Pull request by mikenachbaur-okta**
This permits tokens to be uniquely identified by UUIDs, and to associate custom developer-assigned metadata to retrieve tokens based on their arguments.

### CODE REVIEW

File: Sources/AuthFoundation/Token Management/Internal/KeychainTokenStorage.swift [hunk 1]
```diff
@@ -15,71 +15,47 @@
...
 
     weak var delegate: TokenStorageDelegate?
     
-    private lazy var _defaultTokenId: String? = {
-        if let defaultResult = try? Keychain
-            .Search(account: KeychainTokenStorage.defaultTokenName)
-            .get()
-        {
-            return String(data: defaultResult.value, encoding: .utf8)
+    private(set) lazy var defaultTokenID: UUID? = {
```
Comment (stevelind-okta) [inline Sources/AuthFoundation/Token Management/Internal/KeychainTokenStorage.swift:24]: Might be simpler/more robust to make  it a `String`? UUID is fine of course but can't think of many benefits and you need to keep converting back and forth given that keychain can't store this type.

File: Sources/AuthFoundation/Token Management/Internal/UserDefaultsTokenStorage.swift [hunk 2]
```diff
@@ -24,119 +25,121 @@ class UserDefaultsTokenStorage: TokenStorage {
...
+    private lazy var allTokens: [UUID:Token] = {
+        if let data = userDefaults.data(forKey: UserDefaultsKeys.allTokensKey),
+           let result = try? JSONDecoder().decode([UUID:Token].self, from: data)
+        {
+            return result
+        }
+        return [:]
+    }()
+    
+    private lazy var metadata: [UUID:[String:String]] = {
```
Comment (stevelind-okta) [inline Sources/AuthFoundation/Token Management/Internal/UserDefaultsTokenStorage.swift:47]: nit style guide 
```suggestion
    private lazy var metadata: [UUID: [String: String]] = {
```

File: Sources/AuthFoundation/Token Management/Internal/UserDefaultsTokenStorage.swift [hunk 3]
```diff
@@ -24,119 +25,121 @@ class UserDefaultsTokenStorage: TokenStorage {
...
     
-    func remove(token: Token) throws {
-        guard let index = allTokens.firstIndex(of: token) else { return }
-        allTokens.remove(at: index)
-        
-        if _defaultToken == token {
-            _defaultToken = nil
-            delegate?.token(storage: self, defaultChanged: nil)
-        }
+    func assign(metadata: [String:String], for id: UUID) throws {
```
Comment (stevelind-okta) [inline Sources/AuthFoundation/Token Management/Internal/UserDefaultsTokenStorage.swift:126]: For me had to read this before I understood what it meant
```suggestion
    func assignMetadata(_ metadata: [String:String], for id: UUID) throws {
```

**Review by stevelind-okta — COMMENTED** [review]
Looks good. Couple suggestions:

1. For me the native `UUID` is adding a lot of code and will continue to do so, would prefer `String` type as will be represented if sent to backend etc. Interested to hear your thoughts on the benefits 😄 

2. Style guide on dictionaries `[KeyType: ValueType]`, add a swift linter to this proj so we don't diverge too much? Can use `okta-devices-swift` as a good linter starting point?

File: Sources/AuthFoundation/Token Management/Internal/UserDefaultsTokenStorage.swift (hunk 3 above)
Comment (mikenachbaur-okta) [reply Sources/AuthFoundation/Token Management/Internal/UserDefaultsTokenStorage.swift:126]: Or, I should go back to my original `setMetadata` name...

**Review by laura-rodriguez — APPROVED** [review]

**Review by laurarodriguez-okta — APPROVED** [review]

**Pull request merged**